utildb/gen_deleted_accounts.go

```diff
@@ -59,100 +67,284 @@ func readAccounts(ch chan proxy.ContractLiveliness, deleteHistory *map[common.Ad
 
 	statedb, err = state.MakeOffTheChainStateDB(ss.GetInputState(), tx.Block, state.NewChainConduit(cfg.ChainID == utils.EthereumChainID, utils.GetChainConfig(cfg.ChainID)))
 	if err != nil {
-		return err
+		return nil, err
 	}
 
 	//wrapper
-	statedb = proxy.NewDeletionProxy(statedb, ch, cfg.LogLevel)
+	statedb = proxy.NewDeletionProxy(statedb, ch, log)
 
 	_, err = processor.ProcessTransaction(statedb, int(tx.Block), tx.Transaction, ss)
 	if err != nil {
-		return nil
+		return nil, nil
 	}
 
 	close(ch)
-	des, res := readAccounts(ch, deleteHistory)
-	if len(des)+len(res) > 0 {
-		// if transaction completed successfully, put destroyed accounts
-		// and resurrected accounts to a database
-		if tx.Substate.Result.Status == types.ReceiptStatusSuccessful {
-			err = ddb.SetDestroyedAccounts(tx.Block, tx.Transaction, des, res)
-			if err != nil {
-				return err
-			}
-		}
+
+	livelinessArr := make([]proxy.ContractLiveliness, 0)
+	for liveliness := range ch {
+		livelinessArr = append(livelinessArr, liveliness)
 	}
 
-	return nil
+	return livelinessArr, nil
 }
 
 // GenDeletedAccountsAction replays transactions and record self-destructed accounts and resurrected accounts.
+// Uses round-robin task assignment system to workers to keep order while utilizing parallelism.
 func GenDeletedAccountsAction(cfg *utils.Config, ddb *substate.DestroyedAccountDB, firstBlock uint64, lastBlock uint64) error {
-	var err error
-
-	err = utils.StartCPUProfile(cfg)
+	err := utils.StartCPUProfile(cfg)
 	if err != nil {
 		return err
 	}
 
 	log := logger.NewLogger(cfg.LogLevel, "Generate Deleted Accounts")
-
 	log.Noticef("Generate deleted accounts from block %v to block %v", firstBlock, lastBlock)
 
-	start := time.Now()
-	sec := time.Since(start).Seconds()
-	lastSec := time.Since(start).Seconds()
-	txCount := uint64(0)
-	lastTxCount := uint64(0)
-	var deleteHistory = make(map[common.Address]bool)
+	processor := executor.MakeTxProcessor(cfg)
+
+	wg := sync.WaitGroup{}
+	stopChan := make(chan struct{})
+	errChan := make(chan error)
 
 	iter := substate.NewSubstateIterator(firstBlock, cfg.Workers)
 	defer iter.Release()
 
-	processor := executor.MakeTxProcessor(cfg)
+	//error handling routine
+	encounteredErrors := errorHandler(stopChan, errChan)
+
+	// feeder to send tasks to workers
+	workerInputChannels := taskFeeder(&wg, cfg, iter, lastBlock, stopChan, log)
+
+	// prepare workers to process transactions
+	workerOutputChannels := launchWorkers(&wg, cfg, workerInputChannels, processor, stopChan, errChan, log)
+
+	// collect results from workers and orders them
+	orderedResults := resultCollector(&wg, cfg, workerOutputChannels, stopChan)
+
+	// process ordered txLivelinessResults
+	resolveDeletionsAndResurrections(ddb, orderedResults, stopChan, errChan)
 
-	for iter.Next() {
-		tx := iter.Value()
-		if tx.Block > lastBlock {
-			break
+	// wait until feeder, workers and collector are done
+	wg.Wait()
+
+	// notify error handler to stop
+	close(errChan)
```

Comment:
Error handler should be stopped using the `stopChan` no?

Reply:
nope error handler handles any error that can come trough `stopChan` therefore it should be the one closing it. As there are no tasks to produce error running anymore after `wg.Wait()` it can be notified to stop listening.